### PR TITLE
perf(Search): row-tap cascade kill via row-level live recompute + compiler-stable refs

### PR DESCRIPTION
**Pull request by adhorodyski**
### Explanation of Change

The Search screen ran a parallel double-merge of live Onyx data: `getSections` pulled the snapshot plus ~18 live inputs (`policies`, `cardFeeds`, `allTransactionViolations`, `allReportMetadata`, `reportActions`, `reportAttributesDerivedValue`, …) while each row was already self-hydrating the same data via its own subscriptions. Row taps wrote optimistic primers that fed back into the screen's live deps, invalidating the 28-dep `useMemo`, re-creating the 24-dep `renderItem`, and reconciling every visible row in the tap frame — visible jank in `SPAN_OPEN_REPORT` (16–45ms).

This PR removes the parallel merge and lets each row own its live computation:

- **A1/A2** — Drop `reportActions`, `exportReportActions`, `allReportMetadata`, `reportAttributesDerivedValue` from `<Search>` and `getSections`. `getSections` becomes snapshot-only for these inputs.
- **Capability booleans** — Replace the array-valued `allActions` field on row items with four primitive booleans (`canPay`/`canApprove`/`canSubmit`/`canChangeApprover`) flowing through selection state and bulk-action consumers. Inlined as primitives at row level so the React Compiler tracks each as a value-dep.
- **`useLiveRowCapabilities`** — Row-level live recompute lives in a single hook consumed by `ExpenseReportListItem`, `TransactionGroupListItem`, `TransactionListItem`. Wave C PRs edit the hook, not the three call sites.
- **`useRowSelection`** — Per-row selection read replaces `applySelectionToItem`. The old screen-level selection dep on `renderItem` is gone; `applySelectionToItem` is deleted.

**Measurements** (React DevTools Profiler, 5 trials of warm row-tap, averaged):

| Metric                     | Before | After  | Delta      | Delta %    |
| -------------------------- | ------ | ------ | ---------- | ---------- |
| Total commits              | 76.8   | 60.4   | -16.4      | -21.4%     |
| Total commit duration (ms) | 631.4  | 502.1  | **-129.3** | **-20.5%** |
| Components rendered        | 6880   | 5459   | -1421      | -20.7%     |

PRD: `docs/superpowers/plans/2026-05-20-search-decomposition-prd.md`

### Fixed Issues

$ https://github.com/Expensify/App/issues/92080
PROPOSAL:

### Tests

1. Open Search with a populated expense report list; tap a row → RHP opens without stall, `SPAN_OPEN_REPORT` P90 below baseline
2. Multi-select 5+ rows → bulk Pay/Approve/Submit/Change-Approver enablement reflects live state
3. Open hold-menu on a payable expense → `nonHeldAmount` math reflects PAY availability
4. Switch variants (Chat, Task, Expense flat, Expense grouped, Expense report) → each renders correctly
5. Pusher update for a visible report → row updates without flickering siblings
- [x] Verify that no errors appear in the JS console

### Offline tests

1. Disconnect; open Search → snapshot renders rows correctly
2. Reconnect → rows pick up live data without errors

### QA Steps

Same as tests

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4510fc76bbf5df699a2575bfb49a276af90f3ed7/src/components/LocaleContextProvider.tsx#L80)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4510fc76bbf5df699a2575bfb49a276af90f3ed7/src/components/LocaleContextProvider.tsx#L116-L123)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.ts or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If new assets were added or existing ones were modified, I verified that:
    - [x] The assets are optimized and compressed (for SVG files, run `npm run compress-svg`)
    - [x] The assets load correctly across all supported platforms.
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I added [unit tests](https://github.com/Expensify/App/blob/main/tests/README.md) for any new feature or bug fix in this PR to help automatically prevent regressions in this user flow.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

</details>

### CODE REVIEW

**Comment by adhorodyski** [comment]
@codex review

**Comment by ikevin127** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If new assets were added or existing ones were modified, I verified that:
    - [x] The assets are optimized and compressed (for SVG files, run `npm run compress-svg`)
    - [x] The assets load correctly across all supported platforms.
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos

https://github.com/user-attachments/assets/9e5eda2d-1b42-4de7-b6d5-aa2e599dae96


https://github.com/user-attachments/assets/1a35b457-d039-4244-a519-04bbe5b378b2

File: src/components/Search/SearchSelectionProvider.tsx [hunk 1]
```diff
@@ -281,4 +272,31 @@ function useSyncSelectedReports(data: TransactionListItemType[] | TransactionGro
...
+ */
+function useSelectionCounts(): {selected: number} {
+    const {selectedTransactions} = useSearchSelectionContext();
+    let selected = 0;
+    for (const key in selectedTransactions) {
+        if (selectedTransactions[key]?.isSelected) {
+            selected += 1;
+        }
+    }
+    return {selected};
```
[the remark this answers is not in the document]
Comment (adhorodyski) [reply src/components/Search/SearchSelectionProvider.tsx:290]: The select-all-matching control (`selectAllMatchingItems(true)`) only renders inside `SearchBulkActionsButton`, which is shown when `selected > 0`, and `shouldShowSelectAllMatchingItems` only turns on once all visible items are already selected. So `selectedTransactions` is never empty when `areAllMatchingItemsSelected` flips true. The count also matches the prior `Object.keys(selectedTransactions).length` logic, so there's no regression here.

File: src/components/Search/SearchList/ListItem/useLiveRowCapabilities.ts [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    const [bankAccountList] = useOnyx(ONYXKEYS.BANK_ACCOUNT_LIST);
+
+    if (!enabled || !snapshotData) {
+        return item;
+    }
+
+    const liveActionsArray = liveReportActions ? (Object.values(liveReportActions) as ReportAction[]) : snapshotActions;
+    const liveAllActions = getActions(
+        snapshotData,
+        snapshotData as unknown as OnyxCollection<TransactionViolation[]>,
```
Comment (ikevin127) [inline src/components/Search/SearchList/ListItem/useLiveRowCapabilities.ts]: **⚠️ Unsafe double-cast in `useLiveRowCapabilities.ts`**

**Why it matters**: `snapshotData` is `SearchResults['data'] | undefined`. Casting it through unknown to `OnyxCollection<TransactionViolation[]>` bypasses the type system entirely. If `getActions` ever changes its parameter shape, this will fail silently at runtime.

**Suggested fix**: Use a proper type guard or typed selector.

File: src/components/Search/SearchSelectionProvider.tsx [hunk 3]
```diff
@@ -277,4 +268,31 @@ function useSyncSelectedReports(data: TransactionListItemType[] | TransactionGro
...
+ * the SearchList header migration and land with that consumer (see PRD `useSelectionCounts`).
+ */
+function useSelectionCounts(): {selected: number} {
+    const {selectedTransactions} = useSearchSelectionContext();
+    let selected = 0;
+    for (const key in selectedTransactions) {
+        if (selectedTransactions[key]?.isSelected) {
+            selected += 1;
+        }
+    }
```
Comment (ikevin127) [inline src/components/Search/SearchSelectionProvider.tsx]: **⚠️ `useSelectionCounts` uses `for...in` without `hasOwnProperty` guard**

**Why it matters**: `for...in` iterates over inherited enumerable properties. While `selectedTransactions` is likely a plain object today, any prototype pollution or future polyfill could leak into this count and show incorrect selection counts in the bulk actions bar.

**Suggested fix**:

```ts
for (const key in selectedTransactions) {
    if (Object.prototype.hasOwnProperty.call(selectedTransactions, key) && selectedTransactions[key]?.isSelected) {
        selected += 1;
    }
}
```

Or simpler, use `Object.entries(selectedTransactions)` which is safe and more idiomatic:

```ts
const selected = Object.entries(selectedTransactions).filter(([, value]) => value?.isSelected).length;
```

File: src/components/Search/SearchList/ListItem/useLiveRowCapabilities.ts [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+};
+
+/**
+ * Row-level live merge of action capabilities. Subscribes to per-report
+ * actions/metadata, re-runs `getActions`, and returns either the input item
+ * (unchanged reference) or a fresh copy with refreshed action + canPay/Approve/
+ * Submit/ChangeApprover. The four booleans stay as primitives at the equality
+ * guard so React Compiler keeps downstream consumers memoized.
+ */
+function useLiveRowCapabilities<T extends LiveRowItem>(params: UseLiveRowCapabilitiesParams<T>): T {
```
Comment (ikevin127) [inline src/components/Search/SearchList/ListItem/useLiveRowCapabilities.ts:45]: ⚠️ **`useLiveRowCapabilities` creates **N** individual **Onyx** subscriptions per row**

Each instance of useLiveRowCapabilities subscribes to:
- `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${reportID}`
- `${ONYXKEYS.COLLECTION.REPORT_METADATA}${reportID}`

For 100 visible rows, that's 200 additional **Onyx** subscriptions (vs. the previous single collection-level subscription at the screen level).

**Why it matters**: While the measurements look good for the tested scenario, `react-native-onyx`'s per-key subscription overhead scales differently than a single collection subscription. On lower-end devices with large lists, this could cause memory pressure and jank when scrolling fast (JIT churn from subscribe/unsubscribe).

**Recommended fix / mitigation**: This is an architectural trade-off, but consider adding a JSDoc warning about the subscription count:

```ts
/**
 * ⚠️ Creates 2 Onyx subscriptions per row. For large lists, consider
 *    capping the number of active subscriptions (e.g., via IntersectionObserver
 *    or windowed subscription).
 */
```

**Comment by ikevin127** [comment]
### 📊 Testing

⚠️ Zero tests for the three new hooks

The PR adds:
- `useLiveRowCapabilities.ts` (82 lines of logic)
- `useRowSelection` (8 lines) 
- `useSelectionCounts` (11 lines)

**Removed**: 242-line `applySelectionToItem` test suite with 18 test cases.
**Added**: No new tests for the replacements.

**Why it matters**: `applySelectionToItem` had comprehensive coverage for edge cases like empty reports, partially selected groups, deleted transactions, and reference identity. The new `useRowSelection` replicates some of this but forgoes the transaction-level selection logic (now inline in `ExpenseReportListItem`).

There's no test verifying that selecting all transactions in a report marks the report row as selected, or that the reference-equality optimization in `useLiveRowCapabilities` works.

**Suggested tests to add**:

```ts
// tests/components/Search/useLiveRowCapabilitiesTest.ts
describe('useLiveRowCapabilities', () => {
    it('returns same reference when capabilities have not changed', () => {
        // verify equality guard in useLiveRowCapabilities
    });
    it('updates action when reportActions Onyx key changes', () => {
        // simulate Pusher update
    });
});

// tests/components/Search/useRowSelectionTest.ts
describe('useRowSelection', () => {
    it('marks row selected when areAllMatchingItemsSelected is true', () => {
        // even if key is not in selectedTransactions
    });
    it('reads per-key state when areAllMatchingItemsSelected is false', () => {
        // normal single-key selection
    });
});
```

**Review by ikevin127 — APPROVED** [review]
**🟢 LGTM**

**Merge readiness**: Conditional -> needs the type-safety cast fixed, `for...in` loop hardened, and test coverage for the new hooks before I'd feel confident shipping this. See comments above for details 🙌

**Comment by flaviadefaria** [comment]
No new product considerations - removing my assignment and unsubscribing.

File: src/components/Search/SearchList/ListItem/useLiveRowCapabilities.ts (hunk 2 above)
Comment (adhorodyski) [reply src/components/Search/SearchList/ListItem/useLiveRowCapabilities.ts]: Removed the cast entirely. Added a typed `getViolationsFromSearchData` helper in `SearchUIUtils` that narrows `data` to violation keys via the existing `isViolationEntry` guard — `getActions` now receives a properly-typed `OnyxCollection<TransactionViolation[]>`, no `unknown` hop.

File: src/components/Search/SearchSelectionProvider.tsx (hunk 3 above)
Comment (adhorodyski) [reply src/components/Search/SearchSelectionProvider.tsx]: Done — switched to `Object.values(selectedTransactions).filter((value) => value?.isSelected).length`.

File: src/components/Search/SearchList/ListItem/useLiveRowCapabilities.ts (hunk 4 above)
Comment (adhorodyski) [reply src/components/Search/SearchList/ListItem/useLiveRowCapabilities.ts:45]: This is intentional — it replaces the screen-level collection merge that re-fired every visible row on any row tap (the regression this PR fixes). The list is virtualized, so subscriptions track on-screen rows, not the full result set; net result is the measured -20.5% commit duration. Added a JSDoc note capturing the trade-off and a windowing escape hatch if a future very large list shows churn.

**Comment by adhorodyski** [comment]
@ikevin127 re: testing — added tests for all three new hooks plus the new helper:

- `useRowSelection` — per-key read + the `areAllMatchingItemsSelected` branch you flagged (row selected even when the key is absent from `selectedTransactions`).
- `useSelectionCounts` — counts only `isSelected` entries.
- `useLiveRowCapabilities` — the reference-identity equality guard (same item ref when capabilities are unchanged), the refreshed-capabilities path, and the disabled short-circuit.
- `getViolationsFromSearchData` — the new typed violations helper.

**Comment by mountiny** [comment]
@ikevin127 can you recheck now please?

**Comment by ikevin127** [comment]
@mountiny Yep, all comments were addressed and changes look good for merge 👍

File: src/components/Search/SearchList/ListItem/useLiveRowCapabilities.ts [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+};
+
+type UseLiveRowCapabilitiesParams<T> = {
+    item: T;
+    reportID: string | undefined;
+    itemKey: string;
+    snapshotData: SearchResults['data'] | undefined;
+    snapshotActions: ReportAction[];
+    enabled: boolean;
+};
```
Comment (mountiny) [inline src/components/Search/SearchList/ListItem/useLiveRowCapabilities.ts:29]: docs, should we update the ai skill to make sure types always have docs?

File: src/libs/SearchUIUtils.ts [hunk 6]
```diff
@@ -2438,6 +2448,38 @@ function getAction(allActions: SearchTransactionAction[], primaryActionExclusion
     return allActions.find((action) => action !== CONST.SEARCH.ACTION_TYPES.CHANGE_APPROVER && !primaryActionExclusions.includes(action)) ?? CONST.SEARCH.ACTION_TYPES.VIEW;
 }
 
+/**
+ * Computes the primary action for a row, applying the same primary-action exclusions as `getSections`
+ * (SUBMIT hidden for non-owners; PAY hidden as the primary action for non-reimbursable-only reports).
+ * Used by the row-level live recompute so the live Action button matches the snapshot.
+ */
+function getPrimaryAction(allActions: SearchTransactionAction[], data: OnyxTypes.SearchResults['data'], key: string, currentAccountID: number): SearchTransactionAction {
```
Comment (mountiny) [inline src/libs/SearchUIUtils.ts:2456]: I think this is not unit tested

**Review by mountiny — APPROVED** [review]
only nabs, can you please add those in follow ups?
